File: backend/app/wallet/chains.py
"""Supported chain configuration — future-ready for USDC balances and staking."""

from dataclasses import dataclass


@dataclass(frozen=True)
class ChainConfig:
    key: str
    name: str
    chain_id: int
    native_symbol: str
    usdc_address: str | None = None


SUPPORTED_CHAINS: dict[str, ChainConfig] = {
    "base": ChainConfig(
        key="base",
        name="Base",
        chain_id=8453,
        native_symbol="ETH",
        usdc_address="0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",
    ),
    "polygon": ChainConfig(
        key="polygon",
        name="Polygon",
        chain_id=137,
        native_symbol="MATIC",
        usdc_address="0x3c499c542cEF5E3811e1192ce70d8cC03d5c3359",
    ),
}


def normalize_chain(chain: str) -> str:
    key = chain.strip().lower()
    if key not in SUPPORTED_CHAINS:
        raise ValueError(f"Unsupported chain: {chain}")
    return key


def chain_label(chain: str) -> str:
    return SUPPORTED_CHAINS[normalize_chain(chain)].name


def is_supported_chain(chain: str) -> bool:
    try:
        normalize_chain(chain)
        return True
    except ValueError:
        return False

File: frontend/src/lib/wallet/chains.ts
import { base, polygon, type Chain } from "viem/chains";

export type ScryChainKey = "base" | "polygon";

export type ScryChainConfig = {
  key: ScryChainKey;
  name: string;
  chainId: number;
  viemChain: Chain;
  nativeSymbol: string;
  usdcAddress: string;
  /** Future: enable when USDC balances ship */
  usdcEnabled: boolean;
};

export const SCRY_CHAINS: Record<ScryChainKey, ScryChainConfig> = {
  base: {
    key: "base",
    name: "Base",
    chainId: base.id,
    viemChain: base,
    nativeSymbol: "ETH",
    usdcAddress: "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",
    usdcEnabled: false,
  },
  polygon: {
    key: "polygon",
    name: "Polygon",
    chainId: polygon.id,
    viemChain: polygon,
    nativeSymbol: "MATIC",
    usdcAddress: "0x3c499c542cEF5E3811e1192ce70d8cC03d5c3359",
    usdcEnabled: false,
  },
};

export const SCRY_SUPPORTED_CHAINS = Object.values(SCRY_CHAINS).map((c) => c.viemChain);

export const DEFAULT_CHAIN_KEY: ScryChainKey = "base";

export function chainFromId(chainId: number): ScryChainKey | null {
  for (const chain of Object.values(SCRY_CHAINS)) {
    if (chain.chainId === chainId) return chain.key;
  }
  return null;
}

export function chainConfig(key: ScryChainKey): ScryChainConfig {
  return SCRY_CHAINS[key];
}
